### PR TITLE
Semi supervised

### PR DESCRIPTION
**Pull request by Edouard360**
* Semi supervised:

    * Add `get_indices` and `get_data_loaders` functions, useful for semi_supervised training

    * Correction of the semi-supervised training scheme, where if dataloaders were unbalanced (more labelled samples than unlabelled for instance), the actual loss minimized was wrong. Now the two new training schemes are

        * `train_semi_supervised_jointly` : the `classification_ratio` parameter can trigger a difference between the relative importance of classification vs. reconstruction loss

        * `train_semi_supervised_alternately` :  at the end of each epoch to optimize the ELBO, the classification error is, in turn, optimized.

    * Better naming convention: train/test => labelled/unlabelled

    * optional in SVAEC: `LinearLogRegClassifier` classifier, only updated every epoch (using sklearn). It acts like a NN on the forward pass - ie. for unlabelled samples,  computations are still entirely happening on GPU. This works best with the `train_semi_supervised_alternately` in the caseof the purified dataset populations

* Others:

    * added `get_latent_mean` and `get_latents` functions. (Only `get_latent` before). Still backward compatible, but removes the need to call .cpu().numpy(). Useful for models with more than one hidden stochastic layer.

    * Correct wrong multinomial formula

    * Add `base.py` file to normalize algorithms methods (VAE, VAEC, SVAEC)

### CODE REVIEW

**Comment by jeff-regier** [comment]
Looks good. Will you add a Juptyer notebook to the examples directory (like the one Chenling made) to show how to use the new semi-supervised code?

**Comment by Edouard360** [comment]
Yes, I can do that for the Cortex Dataset

**Comment by jeff-regier** [comment]
```
/home/ubuntu/anaconda3/lib/python3.6/site-packages/matplotlib/__init__.py:1067: UserWarning: Duplicate key in file "/home/ubuntu/.config/matplotlib/matplotlibrc", line #2
  (fname, cnt))
```
You can get rid of this warning by deleting two of the three rows `/home/ubuntu/.config/matplotlib/matplotlibrc`.... not sure why they put them there.

You can get rid of the h5py warning by upgrading it to version 2.8 (through conda or pip).
